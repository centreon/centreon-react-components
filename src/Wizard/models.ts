import { FormikValues, FormikErrors, FormikHelpers } from 'formik';

export interface StepComponentProps {
  disableNextOnSendingRequests: (sendingRequests: Array<boolean>) => void;
}

export interface Step {
  stepName: string;
  validate?: (values: FormikValues) => FormikErrors<FormikValues>;
  validationSchema?;
  Component: (props: StepComponentProps) => JSX.Element;
  hasActionsBar?: boolean;
  skipFormChangeCheck?: boolean;
}

interface ActionsBarLabels {
  labelPrevious: string;
  labelNext: string;
  labelFinish: string;
}

export interface WizardContentProps {
  sendingRequest: boolean;
  step: Step;
  isLastStep: boolean;
  isFirstStep: boolean;
  disableNextOnSendingRequests: (sendingRequests: Array<boolean>) => void;
  goToPreviousStep: () => void;
  goToNextStep: () => void;
  currentStep: number;
  actionsBarLabels: ActionsBarLabels;
}

interface ConfirmDialogLabels {
  labelTitle: string;
  labelMessage: string;
  labelCancel: string;
  labelConfirm: string;
}

export interface WizardProps {
  steps: Array<Step>;
  onSubmit?: (values: FormikValues, bag: FormikHelpers<FormikValues>) => void;
  initialValues?: FormikValues;
  fullHeight?: boolean;
  open: boolean;
  onClose?: () => void;
  width?: 'lg' | 'md' | 'sm' | 'xl' | 'xs' | false;
  confirmDialogLabels?: ConfirmDialogLabels;
  actionsBarLabels?: ActionsBarLabels;
}

export interface ActionsBarProps {
  isLastStep: boolean;
  isFirstStep: boolean;
  goToPreviousStep: () => void;
  goToNextStep: () => void;
  disableActionButtons: boolean;
  isSubmitting: boolean;
  submit: () => void;
  actionsBarLabels: ActionsBarLabels;
}
