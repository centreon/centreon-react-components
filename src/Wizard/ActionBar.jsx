import React from 'react';
import PropTypes from 'prop-types';
import { styled } from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';

const StepButton = styled(Button)({
  color: '#009fdf',
});

function ActionBar(props) {
  const {
    page,
    isLastPage,
    isSubmitting,
    onCancel,
    onPrevious,
    onNext,
    onFinish,
    labelCancel,
    labelPrevious,
    labelNext,
    labelFinish,
  } = props;

  return (
    <Grid container direction="row" justify="space-between" alignItems="center">
      <Grid item>
        {onCancel && (
          <StepButton
            type="button"
            onClick={(event) => onCancel(event, 'cancel')}
          >
            {labelCancel}
          </StepButton>
        )}
      </Grid>

      <Grid item>
        {page > 0 && (
          <StepButton type="button" onClick={onPrevious}>
            {labelPrevious}
          </StepButton>
        )}

        {isLastPage ? (
          <StepButton type="submit" disabled={isSubmitting} onClick={onFinish}>
            {labelFinish}
          </StepButton>
        ) : (
          <StepButton type="submit" onClick={onNext}>
            {labelNext}
          </StepButton>
        )}
      </Grid>
    </Grid>
  );
}

ActionBar.propTypes = {
  page: PropTypes.number,
  isLastPage: PropTypes.bool,
  isSubmitting: PropTypes.bool,
  onCancel: PropTypes.func,
  onPrevious: PropTypes.func,
  onNext: PropTypes.func,
  onFinish: PropTypes.func,
  labelCancel: PropTypes.string,
  labelPrevious: PropTypes.string,
  labelNext: PropTypes.string,
  labelFinish: PropTypes.string,
};

ActionBar.defaultProps = {
  page: 0,
  isLastPage: true,
  isSubmitting: false,
  onCancel: null,
  onPrevious: null,
  onNext: null,
  onFinish: null,
  labelCancel: 'Cancel',
  labelPrevious: 'Previous',
  labelNext: 'Next',
  labelFinish: 'Finish',
};

export default ActionBar;
