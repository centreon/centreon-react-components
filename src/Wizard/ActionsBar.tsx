import * as React from 'react';

import {
  Grid,
  Button,
  Typography,
  CircularProgress,
  makeStyles,
} from '@material-ui/core';

import { ActionsBarProps } from './models';

const useStyles = makeStyles((theme) => ({
  container: {
    position: 'sticky',
    bottom: 0,
    padding: theme.spacing(0, 1),
    borderTop: `1px solid ${theme.palette.grey[300]}`,
  },
  loader: {
    marginLeft: theme.spacing(1),
  },
}));

const ActionsBar = ({
  isFirstStep,
  goToPreviousStep,
  goToNextStep,
  submit,
  disableActionButtons,
  isLastStep,
  isSubmitting,
  actionsBarLabels,
}: ActionsBarProps): JSX.Element => {
  const classes = useStyles();

  const preventEnterKey = (keyEvent) => {
    if ((keyEvent.charCode || keyEvent.keyCode) === 13) {
      keyEvent.preventDefault();
    }
  };

  const { labelFinish, labelNext, labelPrevious } = actionsBarLabels;

  const labelNextFinish = isLastStep ? labelFinish : labelNext;

  return (
    <Grid
      container
      direction="row"
      justify="flex-end"
      alignItems="center"
      className={classes.container}
    >
      <Grid item>
        {!isFirstStep && (
          <Button
            color="primary"
            onClick={goToPreviousStep}
            onKeyPress={preventEnterKey}
            aria-label={labelPrevious}
          >
            <Typography>{labelPrevious}</Typography>
          </Button>
        )}
      </Grid>
      <Grid item>
        <Button
          color="primary"
          onClick={() => (isLastStep ? submit() : goToNextStep())}
          disabled={disableActionButtons}
          onKeyPress={preventEnterKey}
          aria-label={labelNextFinish}
        >
          <Typography>{labelNextFinish}</Typography>
          {isSubmitting && (
            <CircularProgress size={20} className={classes.loader} />
          )}
        </Button>
      </Grid>
    </Grid>
  );
};

export default ActionsBar;
