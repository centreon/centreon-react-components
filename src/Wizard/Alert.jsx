import React from 'react';
import PropTypes from 'prop-types';
import { styled } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogActions from '@material-ui/core/DialogActions';

const ActionButton = styled(Button)({
  color: '#009fdf',
});

function Alert({
  open,
  onCancel,
  onConfirm,
  labelTitle,
  labelMessage,
  labelCancel,
  labelConfirm,
}) {
  return (
    <Dialog open={open}>
      <DialogTitle>{labelTitle}</DialogTitle>
      {labelMessage && (
        <DialogContent>
          <DialogContentText>{labelMessage}</DialogContentText>
        </DialogContent>
      )}
      <DialogActions>
        <ActionButton onClick={onCancel}>{labelCancel}</ActionButton>
        <ActionButton onClick={onConfirm} autoFocus>
          {labelConfirm}
        </ActionButton>
      </DialogActions>
    </Dialog>
  );
}

Alert.propTypes = {
  open: PropTypes.bool.isRequired,
  onCancel: PropTypes.func.isRequired,
  onConfirm: PropTypes.func.isRequired,
  labelTitle: PropTypes.string,
  labelMessage: PropTypes.string,
  labelCancel: PropTypes.string,
  labelConfirm: PropTypes.string,
};

Alert.defaultProps = {
  labelTitle: 'are you sure ?',
  labelMessage: null,
  labelCancel: 'Cancel',
  labelConfirm: 'Confirm',
};

export default Alert;
