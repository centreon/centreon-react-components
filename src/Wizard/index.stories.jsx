import React from 'react';

import Typography from '@material-ui/core/Typography';

import { Wizard, WizardPage as Page } from '..';

export default { title: 'Wizard' };

export const threeSteps = () => (
  <Wizard open>
    <Page label="step 1">
      <Typography variant="h5" align="center">
        Step 1
      </Typography>
    </Page>
    <Page label="step 2">
      <Typography variant="h5" align="center">
        Step 2
      </Typography>
    </Page>
    <Page label="step3">
      <Typography variant="h5" align="center">
        Step 3
      </Typography>
    </Page>
  </Wizard>
);

export const fullHeight = () => (
  <Wizard open fullHeight>
    <Page>
      <Typography variant="h5" align="center">
        Step 1
      </Typography>
    </Page>
    <Page>
      <Typography variant="h5" align="center">
        Step 2
      </Typography>
    </Page>
    <Page>
      <Typography variant="h5" align="center">
        Step 3
      </Typography>
    </Page>
  </Wizard>
);

export const smallWidth = () => (
  <Wizard open width="xs">
    <Page>
      <Typography variant="h5" align="center">
        Step 1
      </Typography>
    </Page>
    <Page>
      <Typography variant="h5" align="center">
        Step 2
      </Typography>
    </Page>
    <Page>
      <Typography variant="h5" align="center">
        Step 3
      </Typography>
    </Page>
  </Wizard>
);

export const withCustomExitAlertLabels = () => (
  <Wizard
    open
    exitConfirmProps={{
      labelTitle: 'Exit wizard ?',
      labelMessage: "Wizard's progress will not be saved",
      labelConfirm: 'Exit',
    }}
  >
    <Page>
      <Typography variant="h5" align="center">
        Step 1
      </Typography>
    </Page>
    <Page>
      <Typography variant="h5" align="center">
        Step 2
      </Typography>
    </Page>
    <Page>
      <Typography variant="h5" align="center">
        Step 3
      </Typography>
    </Page>
  </Wizard>
);
