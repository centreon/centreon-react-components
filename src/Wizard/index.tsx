import * as React from 'react';

import { equals, length, dec, pipe, inc, filter, isEmpty, not } from 'ramda';
import { Formik } from 'formik';

import { Dialog, makeStyles, DialogContent } from '@material-ui/core';

import Confirm from '../Dialog/Confirm';

import { WizardProps } from './models';
import Stepper from './Stepper';
import WizardContent from './WizardContent';

const useStyles = makeStyles((theme) => ({
  fullHeight: {
    height: '100%',
  },
  dialogContent: {
    display: 'flex',
    backgroundColor: theme.palette.grey[100],
    padding: 0,
  },
}));

const actionsBarLabelsDefaultValues = {
  labelPrevious: 'Previous',
  labelNext: 'Next',
  labelFinish: 'Finish',
};

const Wizard = ({
  steps,
  onSubmit = undefined,
  initialValues = {},
  width = 'sm',
  fullHeight = false,
  open,
  onClose = undefined,
  confirmDialogLabels = undefined,
  actionsBarLabels = actionsBarLabelsDefaultValues,
}: WizardProps): JSX.Element => {
  const classes = useStyles();
  const [currentStep, setCurrentStep] = React.useState(0);
  const [sendingRequest, setSendingRequest] = React.useState(false);
  const [openConfirm, setOpenConfirm] = React.useState(false);

  const isLastStep = pipe(dec, equals(currentStep))(length(steps));

  const isFirstStep = equals(currentStep, 0);

  const goToNextStep = () => {
    if (isLastStep) {
      return;
    }
    setCurrentStep(inc(currentStep));
  };

  const goToPreviousStep = () => {
    if (isFirstStep) {
      return;
    }
    setCurrentStep(dec(currentStep));
  };

  const disableNextOnSendingRequests = (sendingRequests) => {
    setSendingRequest(
      pipe(isEmpty, not)(filter(equals(true), sendingRequests)),
    );
  };

  const submit = (values, bag) => {
    if (isLastStep && onSubmit) {
      onSubmit(values, bag);
      return;
    }

    bag.setSubmitting(false);
  };

  const handleClose = (event, reason) => {
    if (equals(reason, 'backdropClick')) {
      setOpenConfirm(true);
      return;
    }
    onClose?.();
  };

  const handleCloseConfirm = (confirm) => {
    setOpenConfirm(false);

    if (!confirm) {
      return;
    }

    onClose?.();
  };

  const { validate, validationSchema } = steps[currentStep];

  return (
    <>
      <Dialog
        maxWidth={width}
        fullWidth
        open={open}
        onClose={handleClose}
        classes={{
          paper: fullHeight ? classes.fullHeight : undefined,
        }}
      >
        <Stepper steps={steps} currentStep={currentStep} />
        <Formik
          initialValues={initialValues}
          validate={validate}
          validationSchema={validationSchema}
          onSubmit={submit}
          validateOnBlur={false}
          validateOnChange
        >
          <DialogContent className={classes.dialogContent}>
            <WizardContent
              step={steps[currentStep]}
              sendingRequest={sendingRequest}
              isLastStep={isLastStep}
              isFirstStep={isFirstStep}
              disableNextOnSendingRequests={disableNextOnSendingRequests}
              goToPreviousStep={goToPreviousStep}
              currentStep={currentStep}
              actionsBarLabels={actionsBarLabels}
              goToNextStep={goToNextStep}
            />
          </DialogContent>
        </Formik>
      </Dialog>
      <Confirm
        open={openConfirm}
        onCancel={() => handleCloseConfirm(false)}
        onConfirm={() => handleCloseConfirm(true)}
        {...confirmDialogLabels}
      />
    </>
  );
};

export default Wizard;
