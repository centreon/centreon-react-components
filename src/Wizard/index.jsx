import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Formik } from 'formik';
import { makeStyles } from '@material-ui/core/styles';
import Dialog from '@material-ui/core/Dialog';
import DialogContent from '@material-ui/core/DialogContent';
import Box from '@material-ui/core/Box';
import Stepper from './Stepper';
import ActionBar from './ActionBar';
import Confirm from '../Dialog/Confirm';

const isReactElement = (element) => {
  if (
    element &&
    element.type &&
    ['object', 'function', 'symbol'].includes(typeof element.type)
  ) {
    return true;
  }

  return false;
};

const cloneElement = (element, props) => {
  const forwardedProps = isReactElement(element) ? props : {};

  return React.cloneElement(element, { ...forwardedProps });
};

const useWizardStyles = makeStyles((theme) => ({
  fullHeight: {
    height: '100%',
  },
  dialogContent: {
    display: 'flex',
    padding: 0,
    backgroundColor: theme.palette.grey[100],
  },
  form: {
    display: 'flex',
    flexDirection: 'column',
    flex: 1,
  },
}));

const Wizard = (props) => {
  const {
    open,
    onClose,
    initialValues,
    onSubmit,
    width,
    fullHeight,
    actionBarProps,
    exitConfirmProps,
    children,
  } = props;
  const classes = useWizardStyles(props);
  const [page, setPage] = useState(0);
  const [values, setValues] = useState(initialValues);
  const [openConfirm, setOpenConfirm] = useState(false);

  const handleClose = (event, reason) => {
    // close wizard without confirmation if it's the first page
    if (page === 0) {
      onClose('cancel');
    } else {
      setOpenConfirm(true);
      onClose(reason);
    }
  };

  const handleCloseConfirm = (confirm) => {
    setOpenConfirm(false);

    if (confirm === true) {
      onClose('cancel');
    }
  };

  const handleNext = (submittedValues) => {
    setPage(Math.min(page + 1, children.length - 1));
    setValues(submittedValues);
  };

  const handlePrevious = () => {
    setPage(Math.max(page - 1, 0));
  };

  const validate = (currentValues) => {
    const activePage = React.Children.toArray(children)[page];

    return activePage.props.validate
      ? activePage.props.validate(currentValues)
      : {};
  };

  const validationSchema = () => {
    const activePage = React.Children.toArray(children)[page];

    return activePage.props.validationSchema
      ? activePage.props.validationSchema
      : null;
  };

  const handleSubmit = (submittedValues, bag) => {
    const isLastPage = page === React.Children.count(children) - 1;

    if (isLastPage && onSubmit) {
      return onSubmit(submittedValues, bag);
    }

    bag.setTouched({});
    bag.setSubmitting(false);
    handleNext(submittedValues);

    return null;
  };

  const activePage = React.Children.toArray(children)[page];
  const isLastPage = page === React.Children.count(children) - 1;

  return (
    <>
      <Dialog
        classes={{ paper: fullHeight ? classes.fullHeight : null }}
        maxWidth={width}
        fullWidth
        open={open}
        onClose={handleClose}
      >
        <Stepper activeStep={page}>{children}</Stepper>
        <DialogContent className={classes.dialogContent}>
          <Formik
            initialValues={values}
            enableReinitialize={false}
            validate={validate}
            validationSchema={validationSchema()}
            onSubmit={handleSubmit}
          >
            {(bag) => (
              <form className={classes.form} onSubmit={bag.handleSubmit}>
                {cloneElement(activePage, {
                  errors: bag.errors,
                  handleBlur: bag.handleBlur,
                  handleChange: bag.handleChange,
                  handleSubmit: bag.handleSubmit,
                  onPrevious: handlePrevious,
                  onNext: handleNext,
                  setFieldTouched: bag.setFieldTouched,
                  setFieldValue: bag.setFieldValue,
                  submitForm: bag.submitForm,
                  touched: bag.touched,
                  values: bag.values,
                })}
                {!activePage.props.noActionBar && (
                  <ActionBar
                    disabledNext={!bag.isValid || bag.isSubmitting}
                    page={page}
                    isLastPage={isLastPage}
                    onPrevious={handlePrevious}
                    actionBarProps={actionBarProps}
                  />
                )}
              </form>
            )}
          </Formik>
        </DialogContent>
      </Dialog>
      <Confirm
        open={openConfirm}
        onCancel={() => handleCloseConfirm(false)}
        onConfirm={() => handleCloseConfirm(true)}
        {...exitConfirmProps}
      />
    </>
  );
};

Wizard.propTypes = {
  open: PropTypes.bool.isRequired,
  onClose: PropTypes.func,
  initialValues: PropTypes.objectOf(PropTypes.any),
  onSubmit: PropTypes.func,
  width: PropTypes.string,
  fullHeight: PropTypes.bool,
  actionBarProps: PropTypes.objectOf(PropTypes.any),
  exitConfirmProps: PropTypes.objectOf(PropTypes.any),
  children: PropTypes.node.isRequired,
};

Wizard.defaultProps = {
  onClose: null,
  initialValues: {},
  onSubmit: null,
  fullHeight: false,
  width: 'md',
  actionBarProps: null,
  exitConfirmProps: null,
};

export const Page = ({ children, ...props }) => (
  <Box height="100%" style={{ overflow: 'auto' }}>
    {React.Children.toArray(children).map((child) => {
      return cloneElement(child, props);
    })}
  </Box>
);

Page.propTypes = {
  label: PropTypes.string,
  children: PropTypes.node,
};

Page.defaultProps = {
  label: null,
  children: null,
};

export default Wizard;
