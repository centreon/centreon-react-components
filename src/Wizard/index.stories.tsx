import * as React from 'react';

import { useFormikContext, FormikErrors, FormikValues } from 'formik';

import { Typography, TextField } from '@material-ui/core';

import Wizard from '.';
import { StepComponentProps } from './models';

export default { title: 'Wizard' };

export const oneStep = (): JSX.Element => (
  <Wizard
    open
    steps={[
      {
        stepName: 'First step',
        skipFormChangeCheck: true,
        Component: (): JSX.Element => (
          <Typography variant="h5" align="center">
            Step 1
          </Typography>
        ),
      },
    ]}
  />
);

export const oneStepWithoutActionsBar = (): JSX.Element => (
  <Wizard
    open
    steps={[
      {
        stepName: 'First step',
        skipFormChangeCheck: true,
        hasActionsBar: false,
        Component: (): JSX.Element => (
          <Typography variant="h5" align="center">
            Step 1
          </Typography>
        ),
      },
    ]}
  />
);

export const threeStepsWithMediumSize = (): JSX.Element => (
  <Wizard
    open
    width="md"
    steps={[
      {
        stepName: 'First step',
        skipFormChangeCheck: true,
        Component: (): JSX.Element => (
          <Typography variant="h5" align="center">
            Step 1
          </Typography>
        ),
      },
      {
        stepName: 'Second step',
        skipFormChangeCheck: true,
        Component: (): JSX.Element => (
          <Typography variant="h5" align="center">
            Step 2
          </Typography>
        ),
      },
      {
        stepName: 'Third step',
        skipFormChangeCheck: true,
        Component: (): JSX.Element => (
          <Typography variant="h5" align="center">
            Step 3
          </Typography>
        ),
      },
    ]}
  />
);

export const threeStepsWithCustomActionsBarLabels = (): JSX.Element => (
  <Wizard
    open
    width="md"
    actionsBarLabels={{
      labelPrevious: 'Previous step',
      labelNext: 'Next step',
      labelFinish: 'Finish wizard',
    }}
    steps={[
      {
        stepName: 'First step',
        skipFormChangeCheck: true,
        Component: (): JSX.Element => (
          <Typography variant="h5" align="center">
            Step 1
          </Typography>
        ),
      },
      {
        stepName: 'Second step',
        skipFormChangeCheck: true,
        Component: (): JSX.Element => (
          <Typography variant="h5" align="center">
            Step 2
          </Typography>
        ),
      },
      {
        stepName: 'Third step',
        skipFormChangeCheck: true,
        Component: (): JSX.Element => (
          <Typography variant="h5" align="center">
            Step 3
          </Typography>
        ),
      },
    ]}
  />
);

export const threeStepsWithFullHeight = (): JSX.Element => (
  <Wizard
    open
    width="md"
    fullHeight
    steps={[
      {
        stepName: 'First step',
        skipFormChangeCheck: true,
        Component: (): JSX.Element => (
          <Typography variant="h5" align="center">
            Step 1
          </Typography>
        ),
      },
      {
        stepName: 'Second step',
        skipFormChangeCheck: true,
        Component: (): JSX.Element => (
          <Typography variant="h5" align="center">
            Step 2
          </Typography>
        ),
      },
      {
        stepName: 'Third step',
        skipFormChangeCheck: true,
        Component: (): JSX.Element => (
          <Typography variant="h5" align="center">
            Step 3
          </Typography>
        ),
      },
    ]}
  />
);

export const threeStepsWithCustomConfirmDialogLabels = (): JSX.Element => (
  <Wizard
    open
    width="md"
    confirmDialogLabels={{
      labelTitle: 'Exit wizard ?',
      labelMessage: "Wizard's progress will not be saved",
      labelConfirm: 'Exit',
      labelCancel: 'Cancel',
    }}
    steps={[
      {
        stepName: 'First step',
        skipFormChangeCheck: true,
        Component: (): JSX.Element => (
          <Typography variant="h5" align="center">
            Step 1
          </Typography>
        ),
      },
      {
        stepName: 'Second step',
        skipFormChangeCheck: true,
        Component: (): JSX.Element => (
          <Typography variant="h5" align="center">
            Step 2
          </Typography>
        ),
      },
      {
        stepName: 'Third step',
        skipFormChangeCheck: true,
        Component: (): JSX.Element => (
          <Typography variant="h5" align="center">
            Step 3
          </Typography>
        ),
      },
    ]}
  />
);

const FirstStep = ({ disableNextOnSendingRequests }: StepComponentProps) => {
  React.useEffect(() => {
    disableNextOnSendingRequests([true, false, true]);
    setTimeout(() => {
      disableNextOnSendingRequests([false, false, false]);
    }, 1500);
  }, []);

  return (
    <Typography variant="h5" align="center">
      Sending request...
    </Typography>
  );
};

export const twoStepsWithSendingRequest = (): JSX.Element => (
  <Wizard
    open
    steps={[
      {
        stepName: 'First step',
        skipFormChangeCheck: true,
        Component: FirstStep,
      },
      {
        stepName: 'Second step',
        skipFormChangeCheck: true,
        Component: (): JSX.Element => (
          <Typography variant="h5" align="center">
            Step 2
          </Typography>
        ),
      },
    ]}
  />
);

interface Values {
  email?: string;
  password?: string;
}

const Form = (): JSX.Element => {
  const [submitted, setSubmitted] = React.useState<boolean>(false);
  return (
    <Wizard
      initialValues={{
        email: '',
        password: '',
      }}
      onSubmit={(values, { setSubmitting }) => {
        setTimeout(() => {
          setSubmitting(false);
          setSubmitted(true);
        }, 500);
      }}
      open
      steps={[
        {
          stepName: 'First Step',
          validate: (values: Values) => {
            const errors: FormikErrors<FormikValues> = {};
            if (!values.email) {
              errors.email = 'Required';
            } else if (
              !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i.test(values.email)
            ) {
              errors.email = 'Invalid email address';
            }
            return errors;
          },
          Component: () => {
            const {
              handleChange,
              handleBlur,
              values,
              errors,
              touched,
            } = useFormikContext<Values>();

            return (
              <TextField
                type="email"
                name="email"
                label="email"
                onChange={handleChange('email')}
                onBlur={handleBlur('email')}
                value={values.email}
                error={!!touched.email && !!errors.email}
                helperText={touched.email ? errors.email : ''}
              />
            );
          },
        },
        {
          stepName: 'Second Step',
          validate: (values: Values) => {
            const errors: FormikErrors<FormikValues> = {};

            if (!values.password) {
              errors.password = 'Required';
            } else if (values.password.length < 6) {
              errors.password = 'Password too short';
            }

            return errors;
          },
          Component: ({ disableNextOnSendingRequests }: StepComponentProps) => {
            const {
              setFieldValue,
              values,
              errors,
              touched,
              handleChange,
              handleBlur,
            } = useFormikContext<Values>();
            React.useEffect(() => {
              if (!values.password) {
                disableNextOnSendingRequests([true, false, true]);
                setTimeout(() => {
                  disableNextOnSendingRequests([false, false, false]);
                  setFieldValue('password', 'pwd');
                }, 1000);
              }
            }, []);

            return (
              <TextField
                type="password"
                name="password"
                label="password"
                onChange={handleChange('password')}
                onBlur={handleBlur('password')}
                value={values.password}
                error={!!touched.password && !!errors.password}
                helperText={touched.password ? errors.password : ''}
              />
            );
          },
        },
        {
          stepName: 'Third Step',
          Component: () => {
            const { values } = useFormikContext();
            return (
              <Typography>
                {!submitted ? JSON.stringify(values) : 'Values submitted'}
              </Typography>
            );
          },
        },
      ]}
    />
  );
};

export const wizardWithInputsAndValidation = (): JSX.Element => <Form />;
