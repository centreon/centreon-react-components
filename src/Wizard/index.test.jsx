import React from 'react';
import { act, render, fireEvent } from '@testing-library/react';
import Typography from '@material-ui/core/Typography';
import Wizard, { Page } from '.';

describe('Wizard', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Wizard open>
        <Page label="step 1">
          <Typography variant="h5" align="center">
            Step 1
          </Typography>
        </Page>
        <Page label="step 2">
          <Typography variant="h5" align="center">
            Step 2
          </Typography>
        </Page>
        <Page label="step3">
          <Typography variant="h5" align="center">
            Step 3
          </Typography>
        </Page>
      </Wizard>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('goes to next and previous steps', async () => {
    const { getByText } = render(
      <Wizard open>
        <Page label="step 1">
          <div>Step 1</div>
        </Page>
        <Page label="step 2">
          <div>Step 2</div>
        </Page>
        <Page label="step3">
          <div>Step 3</div>
        </Page>
      </Wizard>,
    );

    await act(async () => {
      fireEvent.click(getByText('Next').parentNode);
    });

    expect(getByText('Step 2')).toBeInTheDocument();

    await act(async () => {
      fireEvent.click(getByText('Previous').parentNode);
    });

    expect(getByText('Step 1')).toBeInTheDocument();
  });
});
