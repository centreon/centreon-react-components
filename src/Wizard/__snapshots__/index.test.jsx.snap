// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wizard renders correctly 1`] = `null`;
